eclipsesource/theia
# Improve RPC message encoding

#### What it does

- Refactor `RPCMessageEncoder` to enable detection of circular object structures. Circular structures can not be serialized. The detection enables us to fail early instead of indirectly failing once the maximum stack frame limit is reached.
- Add a custom encoder for functions. This ensures that direct function entries in arrays can be serialized.
- Add custom encoder for common collection types (maps, sets). Enables direct use of collection types as rpc call arguments (e.g. logging statements)
- Restore behavior of the old protocol if a log RPC call can not be sent to the backend due to an encoding error 
   - Add custom `EncodingError` type to explicitly catch these errors
   - Refactor `loggerFrontendModule` to catch encoding errors and continue execution normally. i.e. log call is just logged in frontend console and not sent to backend

Fixes https://github.com/eclipse-theia/theia/issues/11249
 
Contributed on behalf of STMicroelectronics
#### How to test

The bug described in https://github.com/eclipse-theia/theia/issues/11249#issuecomment-1147961021 should be fixed. i.e.:

1. Add this log to DebugSessionManager set currentSession:
 ```bash
 console.log('SENTINEL FOR SETTING THE CURRENT SESSION', current);
```

2. Start a debug session inside the application.
3. Observe that the frontend (browser) logs log the message as expected.
4. Observe that no error logged in the frontend console
5. Observe that the backend logs do not show the message
( Same behavior as with the old RPC protocol)
#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
